docs/design_proposals/multi_environment_proposal.md

```diff
@@ -0,0 +1,345 @@
+
+### Feature & Environment Set Definitions
+Introduce environment sets into the `pixi.toml` this describes environments based on `feature`'s. Introduce features into the `pixi.toml` that can describe parts of environments.
+As an environment goes beyond just `dependencies` the `features` should be described including the following fields:
+
+- `dependencies`: The conda package dependencies
+- `pypi-dependencies`: The pypi package dependencies
+- `system-requirements`: The system requirements of the environment
+- `activation`: The activation information for the environment
+- `platforms`: The platforms the environment can be run on.
+- `channels`: The channels used to create the environment.
+- `target`: All the above features but also separated by targets.
+- `tasks`: Feature specific tasks, tasks in one environment are selected as default tasks for the environment.
+
+
+```toml title="Default features" linenums="1"
+[dependencies] # short for [feature.default.dependencies]
+python = "*"
+numpy = "==2.3"
+
+[pypi-dependencies] # short for [feature.default.pypi-dependencies]
+pandas = "*"
+
+[system-requirements] # short for [feature.default.system-requirements]
+libc = "2.33"
+
+[activation] # short for [feature.default.activation]
+scripts = ["activate.sh"]
+```
+
+```toml title="Different dependencies per feature" linenums="1"
+[feature.py39.dependencies]
+python = "~=3.9.0"
+[feature.py310.dependencies]
+python = "~=3.10.0"
+[feature.test.dependencies]
+pytest = "*"
+```
+
+```toml title="Full set of environment modification in one feature" linenums="1"
+[feature.cuda]
+dependencies = {cuda = "x.y.z", cudnn = "12.0"}
+pypi-dependencies = {torch = "1.9.0"}
+platforms = ["linux-64", "osx-arm64"]
+activation = {scripts = ["cuda_activation.sh"]}
+system-requirements = {cuda = "12"}
+channels = ["nvidia"] # Would concat instead of overwrite, so the default channels are still used.
+tasks = { warmup = "python warmup.py" }
+target.osx-arm64 = {dependencies = {mlx = "x.y.z"}}
+```
+
+
+```toml title="Define tasks as defaults of an environment" linenums="1"
+[feature.test.tasks]
+test = "pytest"
+
+[environments]
+test = ["test"]
+
+# `pixi run test` == `pixi run --environments test test`
+```
+
+The environment definition should contain the following fields:
+
+- `features: Vec<Feature>`: The features that are included in the environment set, which is also the default field in the environments.
+- `solve-group: String`: The solve group is used to group environments together at the solve stage.
+This is useful for environments that need to have the same dependencies but might extend them with additional dependencies.
+For instance when testing a production environment with additional test dependencies.
+
+```toml title="Creating environments from features" linenums="1"
+[environments]
+# implicit: default = ["default"]
+default = ["py39"] # implicit: default = ["py39", "default"]
```

Comment:
Is there also a way to exclude the default features? Imagine you are working on a python project, you most likely want to define a base set of dependencies as `[dependencies]` but maybe for linting you want to exclude the base feature. 

Maybe `!default`? To explicitly exclude `default`?

Reply:
See https://github.com/prefix-dev/pixi/pull/584#discussion_r1432675873